### PR TITLE
Add nn-modules base 

### PR DESCRIPTION
**Pull request by nhtlongcs**
This pull request includes base nn-modules:

- nn/datasets (provided create from folder or from list)
- nn/metrics (pixel accuracy) 
- nn/model (mobileunet) 
- nn/learner(supervised learner for segmentation task) 
- independent testing scripts for evaluate the results 
- example training script (runable)

### CODE REVIEW

File: nn/src/datasets/ssdf_datasets.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        image_folder_name="images",
+        extension="png",
+        image_size=(224, 224),
+    )
+
+    print(len(dataset))
+
+    for im, label in dataset:
+        print(label)
+        break
```
Comment (VinhLoiIT) [inline nn/src/datasets/ssdf_datasets.py:126]: I think we should not use an absolute path here even for debugging. We should add another script to test loading instead

File: nn/src/models/segmentation/mobileunet.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+
+if __name__ == "__main__":
+    rand_input = torch.rand((1, 3, 224, 224))
+    model = MobileUnet()
+    print(model)
+    model.eval()
+    with torch.no_grad():
+        output = model(rand_input)
+    print(output.size())
```
Comment (VinhLoiIT) [inline nn/src/models/segmentation/mobileunet.py:227]: Similarly

File: nn/src/test.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    device = get_device()
+
+    transform = [
+        tf.ToTensor(),
+        tf.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+    ]
+
+    dataset = SDataset.from_folder(
+        root="/mnt/c/Users/nhoxs/workspace/ssdf/devtools/nn/data",
```
Comment (VinhLoiIT) [inline nn/src/test.py:57]: Here, too

File: nn/src/train.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    device = get_device()
+
+    transform = [
+        tf.ToTensor(),
+        tf.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+    ]
+    print("aaaaa")
+
+    dataset = SDataset.from_folder(
+        root="/mnt/c/Users/nhoxs/workspace/ssdf/devtools/nn/data",
```
Comment (VinhLoiIT) [inline nn/src/train.py:29]: Here, too

**Review by VinhLoiIT — CHANGES_REQUESTED** [review]
Try updating these paths as suggested.
Also, don't add `nn/requirements.txt` since it is empty.

File: nn/src/test.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    avg_loss = result[0]
+    print("Loss:", avg_loss)
+
+    for k in result[1].keys():
+        result[1][k].summary()
+
+
+if __name__ == "__main__":
+    # test command
+    # python test.py --model /mnt/c/Users/nhoxs/workspace/ssdf/devtools/nn/src/tmp/best_loss.pth --data /mnt/c/Users/nhoxs/workspace/ssdf/devtools/nn/data --img images --msk mask --train
```
Comment (VinhLoiIT) [inline nn/src/test.py:99]: Please move example command to README file

File: nn/src/train.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        criterion=criterion,
+        optimizer=optimizer,
+        save_dir="./tmp",
+    )
+    model_learner.fit(n_epochs=3, metrics=metric)
+
+
+if __name__ == "__main__":
+    # test command
+    # python train.py --model MobileUnet --data /mnt/c/Users/nhoxs/workspace/ssdf/devtools/nn/data --img images --msk mask --train
```
Comment (VinhLoiIT) [inline nn/src/train.py:65]: This too

**Review by VinhLoiIT — CHANGES_REQUESTED** [review]

File: nn/src/datasets/ssdf_datasets.py [hunk 7]
```diff
@@ -47,8 +47,12 @@ def __init__(
             if m_transform is not None
             else tf.Compose([tf.Resize(self.image_size), tf.ToTensor(),])
         )
-        self.list_rgb = rgb_path_ls
-        self.list_mask = mask_path_ls
+        self.list_rgb = rgb_path_ls[:100]
+        self.list_mask = mask_path_ls[:100]
```
Comment (VinhLoiIT) [inline nn/src/datasets/ssdf_datasets.py]: What is this?

File: nn/src/datasets/ssdf_datasets.py [hunk 8]
```diff
@@ -109,14 +111,17 @@ def from_folder(
...
+
     import argparse
 
     parser = argparse.ArgumentParser()
     parser.add_argument("--data-path", type=str, required=True)
     parser.add_argument("--img-folder-name", type=str, required=True)
     parser.add_argument("--msk-folder-name", type=str, required=True)
     parser.add_argument("--train", action="store_true", default=False)
-    parser.add_argument("--extension", default="png", type=str)
+    parser.add_argument("--ex   tension", default="png", type=str)
```
Comment (VinhLoiIT) [inline nn/src/datasets/ssdf_datasets.py]: Please update

File: nn/src/test.py [hunk 9]
```diff
@@ -71,7 +90,7 @@ def test():
 
     dataset = SDataset.from_folder(
         root=args.data_path,
-        train=args.train,
+        train=True,
```
Comment (VinhLoiIT) [inline nn/src/test.py]: why this always set to True?

File: nn/src/utils/segmentation.py [hunk 10]
```diff
@@ -1,9 +1,10 @@
...
 
 
-def multi_class_prediction(output):
+def multi_class_prediction(output: Tensor) -> Tensor:
     return torch.argmax(output, dim=1)
 
 
-def binary_prediction(output, thresh=0.0):
+def binary_prediction(output: Tensor, thresh: int = 0.0) -> Tensor:
     return (output.squeeze(1) > thresh).long()
```
Comment (VinhLoiIT) [inline nn/src/utils/segmentation.py]: If `threshold` is float, then typing should be `float`

**Review by VinhLoiIT — CHANGES_REQUESTED** [review]

**Review by VinhLoiIT — APPROVED** [review]
